gpiozero/internal_devices.py

```diff
@@ -279,45 +286,40 @@ def __init__(self, start_time, end_time, utc=True):
         self._end_time = None
         self._utc = True
         super(TimeOfDay, self).__init__()
-        self.start_time = start_time
-        self.end_time = end_time
-        self.utc = utc
+        if start_time == end_time:
+            raise ValueError('end_time cannot equal start_time')
+        self._start_time = self._validate_time(start_time)
+        self._end_time = self._validate_time(end_time)
+        self._utc = utc
         self._fire_events(self.pin_factory.ticks(), None)
 
     def __repr__(self):
         return '<gpiozero.TimeOfDay active between %s and %s %s>' % (
                 self.start_time, self.end_time, ('local', 'UTC')[self.utc])
 
+    def _validate_time(self, value):
+        if isinstance(value, datetime):
+            print("dt", value)
+            value = value.time()
+        if not isinstance(value, time):
+            raise ValueError(
+                'start_time and end_time must be a datetime, or time instance')
+        return value
+
     @property
     def start_time(self):
         """
         The time of day after which the device will be considered active.
         """
         return self._start_time
 
-    @start_time.setter
-    def start_time(self, value):
-        if isinstance(value, datetime):
-            value = value.time()
-        if not isinstance(value, time):
-            raise ValueError('start_time must be a datetime, or time instance')
-        self._start_time = value
-
     @property
     def end_time(self):
         """
         The time of day after which the device will be considered inactive.
         """
         return self._end_time
 
-    @end_time.setter
```

Comment:
With the removal of the `start_time`, `end_time` and `utc` setters, I guess TimeOfDay objects are now effectively "readonly"?

Reply:
Yes - decided it was easier than validating changes.

Reply:
I guess the only validation needed (in this case) is that start_time != end_time? :man_shrugging: 

Nice catch with the tests that work with the 'active' period crossing the midnight boundary BTW.

Reply:
https://github.com/RPi-Distro/python-gpiozero/blob/b8b7a49d782af56f5f153a187826da3cd2e38a59/gpiozero/internal_devices.py#L289

Reply:
Yeah I saw that, I was just wondering why you couldn't do that in the setters too?

Although now that I think about it, I guess you ought to convert both the start_time and end_time parameters to `time` instances, _before_ checking if they're equal, in case somebody passes in a `time` for one parameter and a `datetime` for another parameter? (and add unit-tests for this too :slightly_smiling_face: )

Reply:
Added tests and fixed!